### PR TITLE
Handle initialization options

### PR DESCRIPTION
**Pull request by ayoub-benali**
fixes https://github.com/scalameta/metals/issues/1614

Migrated from `clientExperimentalCapabilities`
- [x] statusBar
- [x] didFocus
- [x] slowTask
- [x] InputBox
- [x] quickPick
- [x] execute-client-command
- [x] doctorFormat

Migrated from server properties 
- [x] exit-on-shutdown
- [x] completion-item.detail
- [x] completion-item.documentation
- [x] isHoverDocumentationEnabled
- [x] hover.documentation
- [x] signature-help.documentation
- [x] completion-item.resolve
- [x] http

The following have been left out because they are none boolean settings so we have to decide on a priority if setting is set both ways: 
- icons
- glob-syntax
- completionCommand
- signature-help.command
- overrideDefFormat

- [x] Update doc to make initialization options the recommended way for this flags

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/InitializationOptions.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package scala.meta.internal.metals
+import com.google.gson.JsonElement
+import org.eclipse.{lsp4j => l}
+
+final case class InitializationOptions(
+    statusBar: String = "off",
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/InitializationOptions.scala:6]: I'd recommend trying to stay as close to the names we use in `ClientExperimentalCapabilities`.

https://github.com/scalameta/metals/blob/8dc77885d48f5ebd7b97f59b17c6ef850beb7c73/metals/src/main/scala/scala/meta/internal/metals/ClientExperimentalCapabilities.scala#L6-L16

**Comment by ayoub-benali** [comment]
@ckipp01 while I am at it. Do you think other compiler flags likes `isCompletionItemResolve` should be read from `initializationOptions` as well? 
Because I see them set [here](https://github.com/scalameta/metals/blob/0a3ec5b82439b7da7936b770ce6d250dfe8fbd44/metals/src/main/scala/scala/meta/internal/metals/MetalsServerConfig.scala#L136) for example for vim

Basically what in `PresentationCompilerConfig.scala`

**Comment by ckipp01** [comment]
> @ckipp01 while I am at it. Do you think other compiler flags likes `isCompletionItemResolve` should be read from `initializationOptions` as well?
> Because I see them set [here](https://github.com/scalameta/metals/blob/0a3ec5b82439b7da7936b770ce6d250dfe8fbd44/metals/src/main/scala/scala/meta/internal/metals/MetalsServerConfig.scala#L136) for example for vim
> 
> Basically what in `PresentationCompilerConfig.scala`

Yea, I honestly think those can be as well.

**Comment by ckipp01** [comment]
So one thing to think about that maybe @tgodzik or someone else has thoughts on. Some of the ones I question about migrating are ones that deal directly with LSP extensions that we have created since they aren't technically part of LSP, I'm not 100% sure if they should be part of `initializationOptions` or `clientExperimentalCapabilities`. Some of them for example I know that I'll keep in `clientExperimentalCabailities` since they need to be set dynamically. For example in coc-metals a user can turn on or off `metals/status` depending on whether or not they have a statusline integration. Therefore, that needs to remain in `clientExperimentalCapablities` because `initializationOptions` can't be dynamically set in coc.nvim. @gabro I feel like you also have good input on LSP stuff. Any thoughts on this?

**Comment by ayoub-benali** [comment]
@ckipp01 you mean for `coc-nvim` `metals/status` support could be dropped in the middle of an LSP session ?

**Comment by ckipp01** [comment]
> @ckipp01 you mean for `coc-nvim` `metals/status` support could be dropped in the middle of an LSP session ?

Not in the middle of an LSP session, but in the middle of the sessions for example the user could turn it off, which would trigger a server restart with the new setting. set in `clientExperimentalCapabilities`. However if that setting is hardcoded in the `package.json` file as an `initializationOption`, it won't be able to be updated by the user in that way.

**Comment by ayoub-benali** [comment]
> However if that setting is hardcoded in the package.json file as an initializationOption

I see now that this is the core issue. `initializationOption` can only be set there ?

**Comment by ayoub-benali** [comment]
@ckipp01 I updated the field names and the PR description. There are settings I couldn't migrate because they aren't Boolean so we would need to prioritize their initializationOption value over the server settings, or similar.

Please let me know what you think so I can work on the doc once the settings settled.

File: metals/src/main/scala/scala/meta/internal/metals/Compilers.scala [hunk 2]
```diff
@@ -357,12 +357,18 @@ class Compilers(
...
       .withScheduledExecutorService(sh)
       .withConfiguration(
-        config.compilers.copy(
-          _symbolPrefixes = userConfig().symbolPrefixes,
-          isCompletionSnippetsEnabled =
-            initializeParams.supportsCompletionSnippets,
-          isFoldOnlyLines = initializeParams.foldOnlyLines
-        )
+        initializeParams
+          .map(p => {
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/Compilers.scala:361]:
```suggestion
          .map(params => {
```

File: metals/src/main/scala/scala/meta/internal/metals/ConfiguredLanguageClient.scala [hunk 3]
```diff
@@ -85,7 +92,7 @@ final class ConfiguredLanguageClient(
   override def metalsExecuteClientCommand(
       params: ExecuteCommandParams
   ): Unit = {
-    if (config.executeClientCommand.isOn || clientCapabilities.executeClientCommandProvider) {
+    if (config.executeClientCommand.isOn || clientCapabilities.executeClientCommandProvider || initializationOptions.executeClientCommandProvider) {
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/ConfiguredLanguageClient.scala:95]: In a lot of these boolean checks can you pull them out into a variable like you did below? For example this one being like `isExecuteClientCommandProvider` or something like that. Now that we have three different boolean checks in here, it's getting a bit messy.

**Comment by ckipp01** [comment]
> I see now that this is the core issue. initializationOption can only be set there ?

I'm not 100% sure to be honest. I need to look more into it.

## A bit of context

Just for the others reviewing this, I'll give a bit more backstory to why this pr exists. Initially we set everything with properties. A While back [here](https://github.com/scalameta/metals/pull/1414) I had a pr where I moved a chunk of stuff from server properties to `clientExperimentalCapablities`. In hindsight I think I should have moved them to `initializationOptions` rather than `clientExperimentalCapablities`. Both can be used in the same way to really set any value, but if you look at the [lsp spec](https://microsoft.github.io/language-server-protocol/specification#initialize) you see that the way `clientExperimentalCapablities` is talked about is like the following:

> The experimental can be used to pass experimental capabilities under development.

I think the things we are passing in there aren't really _experimental_ so I question the correctness of having them there. Looking around at some other LSP integrations, I've noticed that many make it somewhat easy to set `initializationOptions`, but not so easy to set `clientExperimentalCapablities`. I found this the case with `kak-lsp` when looking into Kakuaone.

I'm propose that we communicate that we are moving away from things being in `clientExperimentalCapablities` (maybe even deprecate the values we have in there if that makes sense?) and document that the way to set values should be in `initializationOptions` rather than `clientExperimentalCapablities` unless like the spec says, they are in development and are actually experimental.

In theory, we could set everything we need to this way and completely move away from having to even set a client as a property. I think this would be a good goal.

@ayoub-benali thanks a lot for working on this.

@kurnevsky I'm also tagging you on here because you were involved with the emacs fold integrating this stuff as `clientExperimentalCapablities`. Just want to give you a heads up on this conversation as it's a direction we are discussing moving to.

**Comment by ayoub-benali** [comment]
> Looking around at some other LSP integrations, I've noticed that many make it somewhat easy to set initializationOptions, but not so easy to set clientExperimentalCapablities.

It was exactly the case for sublime LSP package, I had to add it https://github.com/sublimelsp/LSP/pull/971 this might also happen for other clients.

**Comment by ayoub-benali** [comment]
> Though I think it would make sense to have an update method inside MetalsServerConfig and we could have all the logic for resolving conflicts in there

Yes but the biggest issue is that `config` in `MetalsLanguageServer` is immutable while `clientExperimentalCapabilities` and `initializationOptions` are mutables. They get initialized with default values and then later get mutated once the initialization request is done. But in the mean time their references are passed around ...

A proper solution where all these settings would get merged in a single place would require quite some refactoring of `MetalsLanguageServer`

File: docs/editors/new-editor.md [hunk 4]
```diff
@@ -78,7 +78,7 @@ example, in vim-lsc the `window/logMessage` notification is always displayed in
 the UI so `-Dmetals.status-bar=log-message` can be configured to direct
 higher-priority messages to the logs. However, whenever possible, if the client
 supports the ability to add in `experimental` items to the `ClientCapabilities`
-interface, this is preferable.
+interface and/or to `InitializationOptions`, this is preferable.
```
Comment (ckipp01) [inline docs/editors/new-editor.md:81]:
```suggestion
interface or to `InitializationOptions`, this is preferable.
```
I'd say just use **or**. We don't want confusion over whether they should set them both or not.

File: metals/src/main/scala/scala/meta/internal/metals/ConfiguredLanguageClient.scala [hunk 5]
```diff
@@ -25,22 +25,31 @@ final class ConfiguredLanguageClient(
...
-    if (config.statusBar.isOn || clientCapabilities.statusBarIsOn) {
+    val statusBarIsOn = config.statusBar.isOn ||
+      clientCapabilities.statusBarIsOn ||
+      initializationOptions.statusBarIsOn
+
+    if (statusBarIsOn) {
       underlying.metalsStatus(params)
     } else if (params.text.nonEmpty && !pendingShowMessage.get()) {
-      if (config.statusBar.isShowMessage || clientCapabilities.statusBarIsShowMessage) {
+      if (config.statusBar.isShowMessage || clientCapabilities.statusBarIsShowMessage || initializationOptions.statusBarIsShowMessage) {
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/ConfiguredLanguageClient.scala:50]: Nit: Can you do the same for this one and the one below and separate this out into a variable like you did for `statusBarIsOn`. Then we can just do checks on `statusBarIsShow` and `statusBarIsLog`.

File: docs/editors/new-editor.md (hunk 4 above)
Comment (ayoub-benali) [reply docs/editors/new-editor.md:81]: I was trying to say that some setting should be part of `InitializationOptions` other in `experimental`

Comment (ckipp01) [reply docs/editors/new-editor.md]: Understood 👍

**Pull request merged**